Fix false negatives in db::apply callable check

nilsvu opened the pull request:
## Proposed changes

The checks reported failures in cases where actually calling the
function would have succeeded, because they missed a `const &`.
Using `decltype()` now to guarantee the argument type is correct.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Fix a typo in the docs

nilsdeppe commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1200,26 +1200,19 @@ static constexpr auto apply(F&& f, const DataBox<BoxTags>& box,
-        std::remove_pointer_t<F>,
-        tmpl::conditional_t<std::is_same_v<ArgumentTags, ::Tags::DataBox>,
-                            const DataBox<BoxTags>&,
-                            const_item_type<ArgumentTags, BoxTags>>...,
```
I think the bug is in `const_item_type` because it should be returning a `const T&` but returns a `T`. This is because of the `std::decay_t` on the outside. Why not instead fix that instead? I think this is the reason why we have (but shouldn't need) `const const_item_type<A, B>&` is so many places in stead of `const_item_type<A, B>`.

Note that @markscheel is also fixing another bug in `const_item_type`. The bug there is that because of the use of `decltype` and `declval` a complete type is needed when it shouldn't be. The fix for both basically amounts to:

```cpp
template <typename T>
struct ConvertToConst {
  using type = std::add_const_t<std::add_lvalue_reference_t<T>>;
  // or
  // using type = const T&;
};

template <typename T>
struct ConvertToConst<std::unique_ptr<T>> {
  using type = std::add_const_t<std::add_lvalue_reference_t<T>>;
  // or
  // using type = const T&;
};

template <typename Tag, typename TagList = NoSuchType>
using const_item_type =
    typename ConvertToConst<std::decay_t<storage_type<Tag, TagList>>>;
```
I don't think the interior `std::decay_t` on the last line is needed either, but I'm not sure.

nilsvu replied:
Rewrote the implementation to your suggestions 👍

markscheel replied:
Oh, I just saw this.  This partially fixes the same problem as my #2726.  We should figure out how to combine both PRs.  I basically have the same changes as you in DataBoxTag.hpp but other changes elsewhere...

nilsvu replied:
@markscheel could you remove the `std::decay_t` and take over the `std::optional<std::reference_wrapper>` test in your PR? Then I can close this one.

markscheel replied:
@nilsleiffischer OK I can remove my `std::decay_t`. Do you still want all of your changes in all the other files besides DataBoxTag.hpp?

markscheel replied:
Wait, I get tons of errors when I remove my `std::decay_t`. Maybe I remove my DataBoxTag.hpp changes from my PR and keep the other ones, and you merge your PR?

nilsvu replied:
When you remove `std::decay_t` you have to go through the uses of `const_item_type` and remove the `const &` they are wrapped in. I can also do that later, but could be easier if you include that in your PR that touches the files anyway 🤷‍♂️

markscheel replied:
But didn't you already remove all the `const &` stuff in your PR and add the tests?  It's probably easier for me to just remove one commit (my DataBoxTag.hpp) from my PR and merge in mine after yours.

nilsvu replied:
It took me maybe 5 minutes :D But sure, I'm happy with just merging this PR quickly and your PR building on it. Could you give it a review then please?

nilsvu committed:
Fix false negatives in db::apply callable check

The callable-checks in db::apply reported failures in cases where
actually calling the function would have succeeded, because they
missed a `const &`. This commit includes the `const &` in the
`const_item_type` helper metafunction.

Also, by removing the use of `decltype` and `declval` in
`const_item_type` it doesn't require a complete type anymore.

markscheel commented on tests/Unit/DataStructures/DataBox/Test_DataBox.cpp:
```diff
@@ -703,6 +705,17 @@ void test_apply() noexcept {
+        [](const std::optional<
+            std::reference_wrapper<const std::vector<double>>>
+               optional_ref) {
+          CHECK(optional_ref.has_value());
```
Do you want to check whether the thing referred to has the correct value?

markscheel reviewed (commented):
Just one small suggestion, testing the value of the vector.

nilsvu commented:
@markscheel slightly adjusted the test to check the expected value, please take another look

nilsvu commented:
@nilsdeppe @kidder please take another look

nilsdeppe commented:
Thanks for fixing this!
